File: frameworks/native/libs/nvcpud_client/NvCpuClient.cpp
/*
 * Copyright (C) 2011-2012 The Android Open Source Project
 * Copyright (c) 2012, NVIDIA CORPORATION.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <nvcpud/NvCpuClient.h>

#include <binder/IServiceManager.h>

#undef LOG_TAG
#define LOG_TAG "NvCpuClient"

namespace android {

#define RATE_LIMIT_INIT(func) \
    last##func(0),

NvCpuClient::NvCpuClient() :
    failedConnectionAttempts(0),
    NVCPU_FUNC_LIST(RATE_LIMIT_INIT)
    lastConnectionAttempt(0) {}

#undef RATE_LIMIT_INIT

const nsecs_t NvCpuClient::MINIMUM_POKE_WAIT_NS = 40 * 1000000LL;
const nsecs_t NvCpuClient::MINIMUM_CONNECTION_DELAY_NS = 5000 * 1000000LL;
const int NvCpuClient::MAXIMUM_CONNECTION_ATTEMPTS = 100;

status_t NvCpuClient::tryBindToService() {
    sp<IServiceManager> sm = defaultServiceManager();
    if (sm == 0)
        return FAILED_TRANSACTION;
    sp<IBinder> binder = sm->checkService(String16("NvCpuService"));
    if (binder == 0)
        return NAME_NOT_FOUND;
    mInterface = INvCpuService::asInterface(binder);
    if (mInterface == 0)
        return BAD_VALUE;
    LOGI("Successfully bound to service");
    return NO_ERROR;
}

status_t NvCpuClient::bindToService() {
    sp<IServiceManager> sm = defaultServiceManager();
    if (sm == 0)
        return FAILED_TRANSACTION;
    sp<IBinder> binder = sm->getService(String16("NvCpuService"));
    if (binder == 0)
        return NAME_NOT_FOUND;
    mInterface = INvCpuService::asInterface(binder);
    if (mInterface == 0)
        return BAD_VALUE;
    LOGI("Successfully bound to service");
    return NO_ERROR;
}

void NvCpuClient::checkAndTryReconnect(nsecs_t now) {
    if (mInterface != 0)
        return;
    if (failedConnectionAttempts > MAXIMUM_CONNECTION_ATTEMPTS)
        return;
    if (now < lastConnectionAttempt + MINIMUM_CONNECTION_DELAY_NS)
        return;
    lastConnectionAttempt = now;
    status_t res = tryBindToService();
    if (res != NO_ERROR) {
        LOGW("Failed to bind to service");
        failedConnectionAttempts++;
    }
}

bool NvCpuClient::rateLimitAndCheckConnection(nsecs_t* last, nsecs_t now) {
    if (now - *last < MINIMUM_POKE_WAIT_NS)
        return false;
    *last = now;
    checkAndTryReconnect(now);
    if (mInterface == 0)
        return false;
    return true;
}

#define TIMEOUT_DEF(func) \
void NvCpuClient::func(int arg, nsecs_t timeoutNs, nsecs_t now) { \
    if (!rateLimitAndCheckConnection(&last##func, now)) \
        return; \
    status_t res = mInterface->func(arg, timeoutNs); \
    if (res != NO_ERROR) \
        breakConnection(); \
}

#define HANDLE_DEF(func) \
int NvCpuClient::func##Handle(int arg, nsecs_t now) { \
    checkAndTryReconnect(now); \
    if (mInterface == 0) \
        return -1; \
    return mInterface->func##Handle(arg); \
}

NVCPU_FUNC_LIST(TIMEOUT_DEF)
NVCPU_FUNC_LIST(HANDLE_DEF)

#define SET_DEF(func) \
void NvCpuClient::set##func(int arg) { \
    checkAndTryReconnect(systemTime(SYSTEM_TIME_MONOTONIC)); \
    if (mInterface == 0) { \
        return; \
    } \
    mInterface->set##func(arg); \
}

#define GET_DEF(func) \
int NvCpuClient::get##func(void) { \
    checkAndTryReconnect(systemTime(SYSTEM_TIME_MONOTONIC)); \
    if (mInterface == 0) { \
        return -1; \
    } \
  return mInterface->get##func(); \
}

NVCPU_AP_FUNC_LIST(SET_DEF)
NVCPU_AP_FUNC_LIST(GET_DEF)

#undef TIMEOUT_DEF
#undef HANDLE_DEF
#undef SET_DEF
#undef GET_DEF

void NvCpuClient::breakConnection() {
    LOGW("Lost connection to service");
    mInterface = NULL;
}

};

File: frameworks/base/core/java/com/nvidia/NvCpuClient.java
/*
 * Copyright (c) 2012-2013 NVIDIA Corporation.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Class structure based upon Camera in Camera.java:
 * Copyright (C) 2009 The Android Open Source Project
 */

/**
 * @file
 * <b>NVIDIA Tegra Android CPU management</b>
 *
 * @b Description: Exposes management of Tegra cores to speed up
 *    computation or save power when necessary
 */

package com.nvidia;
import android.util.Log;
import java.io.FileDescriptor;

/**
 * @hide
 *
 */
public class NvCpuClient extends Object implements NvCpuClientConstants {
    private static final String TAG = "NvCpuClient";

    /** requires permission ACCESS_DEVICE_POWER or NVIDIA_CPU_POWER
     * fails silently if it's missing;
     * the access check is performed asynchronously
     */

    /*
     * alias for requestCpuFreqMin
     */
    public native void pokeCPU(
            int strength,
            long timeoutNs,
            long nowNanoTime);

    /*
     * Timeout Methods
     *
     * These place a ceiling/floor on system power settings
     * for the duration of timeoutNs.  nowNanoTime is the current
     * time as reported by System.nanoTime, and is used for API
     * rate-limiting
     *
     */
    public native void requestMinOnlineCpuCount(
            int numCpus,
            long timeoutNs,
            long nowNanoTime);

    public native void requestMaxOnlineCpuCount(
            int numCpus,
            long timeoutNs,
            long nowNanoTime);

    public native void requestCpuFreqMin(
            int strength,
            long timeoutNs,
            long nowNanoTime);

    public native void requestCpuFreqMax(
            int strength,
            long timeoutNs,
            long nowNanoTime);

    public native void requestMaxCbusFreq(
            int frequency,
            long timeoutNs,
            long nowNanoTime);

    public native void requestMinCbusFreq(
            int frequency,
            long timeoutNs,
            long nowNanoTime);

    /*
     * Handle methods
     *
     * These return a FileDescriptor that acts as a handle for
     * an open system power request.  Closing the file descriptor
     * ends the requests; it is also closed if the client process dies.
     *
     * nowNanoTIme is the current time as reported by System.nanoTime,
     * it is used for API rate limiting
     */

    public native FileDescriptor requestMinOnlineCpuCountHandle(
            int numCpus,
            long nowNanoTime);

    public native FileDescriptor requestMaxOnlineCpuCountHandle(
            int numCpus,
            long nowNanoTime);

    public native FileDescriptor requestCpuFreqMinHandle(
            int strength,
            long nowNanoTime);

    public native FileDescriptor requestCpuFreqMaxHandle(
            int strength,
            long nowNanoTime);

    public native FileDescriptor requestMaxCbusFreqHandle(
            int frequency,
            long nowNanoTime);

    public native FileDescriptor requestMinCbusFreqHandle(
            int frequency,
            long nowNanoTime);

    /*
     * App Profile methods
     * These methods sets/gets information from sysfs knobs
     * and are used by ActivityStack to modify performance
     * depending on App currently running
     */
    public native void setCpuCoreBias(int bias);

    public native int getCpuCoreBias();

    public native void setGpuCbusCapLevel(int freq);

    public native int getGpuCbusCapLevel();

    public native void setGpuCbusCapState(int state);

    public native int getGpuCbusCapState();

    public native void setCpuHighFreqMinDelay(int delay);

    public native int getCpuHighFreqMinDelay();

    public native void setCpuMaxNormalFreq(int freq);

    public native int getCpuMaxNormalFreq();

    public native void setCpuScalingMinFreq(int freq);

    public native int getCpuScalingMinFreq();

    public native void setGpuScalingEnable(int val);

    public native int getGpuScalingEnable();

    public native void setEnableAppProfiles(int val);

    public native int getEnableAppProfiles();

    public NvCpuClient() {
        mNativeNvCpuClient = 0;
        init();
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            super.finalize();
        } finally {
            if (mNativeNvCpuClient != 0) {
                release();
            }
        }
    }

    private native void init();
    private native void release();
    //This is 32-bit dependent code!
    //

    private native static void nativeClassInit();
    static { nativeClassInit(); }

    private int mNativeNvCpuClient;
}
